File: main/store/shared-state.ts
/**
 * SharedStateCenter — 显式共享状态中心
 *
 * 替代点对点路由，实现 A2A "扔字条"式解耦通信：
 * - TaskQueue: 任何 agent 都可以向某个 agent 的收件箱「扔任务字条」
 * - NotifyQueue: 任务完成/状态变更等通知广播
 * - 接收者主动轮询（pollInbox）获取自己的邮件
 * - 完全解决耦：发布者不知道接收者是否存在、何时消费
 */

import { v4 as uuid } from 'uuid'
import { persistence } from './persistence'

export type TaskQueueStatus = 'pending' | 'claimed' | 'completed' | 'failed'

export interface TaskQueueItem {
  id: string
  group_id: string
  sender_id: string
  receiver_id: string     // 目标 agent id
  content: string
  data?: Record<string, unknown>
  created_at: string
  status: TaskQueueStatus
  claimed_by?: string     // instance_id
  result?: string
  result_data?: Record<string, unknown>
  completed_at?: string
}

export type NotifyType = 'task_complete' | 'task_failed' | 'agent_reply' | 'task_log' | 'coordinator_reply'

export interface NotifyQueueItem {
  id: string
  group_id: string
  type: NotifyType
  sender_id: string
  receiver_id: string     // 'broadcast' 或具体 agent id
  content: string
  data?: Record<string, unknown>
  created_at: string
}

/** 一个群组的队列快照（用于持久化和恢复） */
export interface GroupQueueSnapshot {
  group_id: string
  tasks: TaskQueueItem[]
  notifies: NotifyQueueItem[]
}

class SharedStateCenter {
  /** groupId -> TaskQueueItem[] */
  private taskQueues = new Map<string, TaskQueueItem[]>()

  /** groupId -> NotifyQueueItem[] */
  private notifyQueues = new Map<string, NotifyQueueItem[]>()

  /** 每个 agent 上次轮询时间（用于增量拉取） */
  private lastPollMap = new Map<string, string>()

  // ── 任务队列 ──────────────────────────────────────────────

  /**
   * 向指定 agent 的收件箱「扔任务字条」
   * 发送者无需知道接收者是否在线
   */
  pushTask(params: {
    group_id: string
    sender_id: string
    receiver_id: string
    content: string
    data?: Record<string, unknown>
  }): TaskQueueItem {
    const item: TaskQueueItem = {
      id: uuid(),
      group_id: params.group_id,
      sender_id: params.sender_id,
      receiver_id: params.receiver_id,
      content: params.content,
      data: params.data,
      created_at: new Date().toISOString(),
      status: 'pending',
    }
    const q = this._getTaskQueue(params.group_id)
    q.push(item)
    this._schedulePersist(params.group_id)
    return item
  }

  /**
   * Agent 主动来取一个属于自己的 pending 任务
   * 返回后自动标记为 claimed
   */
  claimTask(groupId: string, agentId: string, instanceId: string): TaskQueueItem | undefined {
    const q = this._getTaskQueue(groupId)
    const idx = q.findIndex(
      t => t.receiver_id === agentId && t.status === 'pending',
    )
    if (idx === -1) return undefined

    const item = q[idx]
    item.status = 'claimed'
    item.claimed_by = instanceId
    q[idx] = item
    this._schedulePersist(groupId)
    return item
  }

  /**
   * Agent 完成任务后写回结果
   * 同时自动在 NotifyQueue 发布 task_complete / task_failed
   */
  completeTask(
    taskId: string,
    success: boolean,
    result?: string,
    resultData?: Record<string, unknown>,
  ): TaskQueueItem | undefined {
    for (const [groupId, q] of this.taskQueues) {
      const item = q.find(t => t.id === taskId)
      if (item) {
        item.status = success ? 'completed' : 'failed'
        item.result = result
        item.result_data = resultData
        item.completed_at = new Date().toISOString()
        this._schedulePersist(groupId)

        // 自动发布通知给 sender
        this.pushNotify({
          group_id: item.group_id,
          type: success ? 'task_complete' : 'task_failed',
          sender_id: item.receiver_id,
          receiver_id: item.sender_id,
          content: result || (success ? '任务完成' : '任务失败'),
          data: { task_id: taskId, ...resultData },
        })
        return item
      }
    }
    return undefined
  }

  /**
   * 查询某个群的所有任务（用于 coordinator monitor）
   */
  listTasks(groupId: string): TaskQueueItem[] {
    return [...this._getTaskQueue(groupId)]
  }

  /**
   * 查询某个 agent 的 pending 任务数量
   */
  countPending(groupId: string, agentId: string): number {
    return this._getTaskQueue(groupId).filter(
      t => t.receiver_id === agentId && t.status === 'pending',
    ).length
  }

  // ── 通知队列 ──────────────────────────────────────────────

  /**
   * 向共享通知队列投递通知
   * receiver_id = 'broadcast' 时会被所有轮询者收到
   */
  pushNotify(params: {
    group_id: string
    type: NotifyType
    sender_id: string
    receiver_id: string
    content: string
    data?: Record<string, unknown>
  }): NotifyQueueItem {
    const item: NotifyQueueItem = {
      id: uuid(),
      group_id: params.group_id,
      type: params.type,
      sender_id: params.sender_id,
      receiver_id: params.receiver_id,
      content: params.content,
      data: params.data,
      created_at: new Date().toISOString(),
    }
    const q = this._getNotifyQueue(params.group_id)
    q.push(item)

    // 通知队列过长时截断（保留最近 500 条）
    if (q.length > 500) {
      q.splice(0, q.length - 500)
    }

    this._schedulePersist(params.group_id)
    return item
  }

  // ── 收件箱轮询（核心解耦 API）──────────────────────────────

  /**
   * 轮询某个 agent 的收件箱：返回属于它的 pending 任务 + 定向/广播通知
   *
   * 这是 AgentEngine 主循环里替代 `inbox.shift()` 的核心方法。
   */
  pollInbox(
    groupId: string,
    agentId: string,
    options?: {
      since?: string       // ISO timestamp，只返回大于此时间的
      taskOnly?: boolean   // 只取任务（不取通知）
    },
  ): { tasks: TaskQueueItem[]; notifies: NotifyQueueItem[] } {
    const since = options?.since
    const key = `${groupId}:${agentId}`

    // 属于我的 pending 任务
    const tasks = this._getTaskQueue(groupId).filter(
      t => t.receiver_id === agentId && t.status === 'pending',
    )

    // 筛选通知：定向给我 或 broadcast，且时间 > since
    const allNotifies = this._getNotifyQueue(groupId)
    const notifies = allNotifies.filter(n => {
      if (n.receiver_id !== 'broadcast' && n.receiver_id !== agentId) return false
      if (!since) return true
      return n.created_at > since
    })

    // 更新时间戳
    this.lastPollMap.set(key, new Date().toISOString())

    return { tasks, notifies }
  }

  /**
   * 一次性拉取 inbox 中最老的一条，用于 AgentEngine 逐个消费
   * 优先返回 tasks，没有 tasks 再返回 notifies
   */
  pollOne(groupId: string, agentId: string):
    | { kind: 'task'; item: TaskQueueItem }
    | { kind: 'notify'; item: NotifyQueueItem }
    | undefined {
    const { tasks, notifies } = this.pollInbox(groupId, agentId)

    if (tasks.length) {
      // 返回最老的一条任务
      const oldest = tasks.sort((a, b) => a.created_at.localeCompare(b.created_at))[0]
      return { kind: 'task', item: oldest }
    }

    if (notifies.length) {
      // 返回最老的一条通知
      const oldest = notifies.sort((a, b) => a.created_at.localeCompare(b.created_at))[0]
      return { kind: 'notify', item: oldest }
    }

    return undefined
  }

  // ── 持久化 ────────────────────────────────────────────────

  /**
   * 启动时从 JSON 恢复所有队列
   */
  async loadAll(): Promise<void> {
    const snapshots = await persistence.loadAllQueues()
    for (const snap of snapshots) {
      this.taskQueues.set(snap.group_id, snap.tasks || [])
      this.notifyQueues.set(snap.group_id, snap.notifies || [])
    }
  }

  private _schedulePersist(groupId: string): void {
    const snap: GroupQueueSnapshot = {
      group_id: groupId,
      tasks: this._getTaskQueue(groupId),
      notifies: this._getNotifyQueue(groupId),
    }
    persistence.scheduleSaveQueue(groupId, snap)
  }

  // ── 内部辅助 ──────────────────────────────────────────────

  private _getTaskQueue(groupId: string): TaskQueueItem[] {
    if (!this.taskQueues.has(groupId)) {
      this.taskQueues.set(groupId, [])
    }
    return this.taskQueues.get(groupId)!
  }

  private _getNotifyQueue(groupId: string): NotifyQueueItem[] {
    if (!this.notifyQueues.has(groupId)) {
      this.notifyQueues.set(groupId, [])
    }
    return this.notifyQueues.get(groupId)!
  }
}

export const sharedState = new SharedStateCenter()

File: src-tauri/src/store/shared_state.rs
//! SharedStateCenter —— A2A 共享状态中心
//!
//! 对应 TS `main/store/shared-state.ts`：
//! - TaskQueue: 向某个 agent 收件箱「扔任务字条」
//! - NotifyQueue: 任务完成/状态变更通知
//! - pollInbox: 接收者主动轮询取信
//!
//! 真消息驱动改造：每个 (group, agent) 对应一个 tokio mpsc 收件箱 channel，
//! push_* 直接 send，引擎端 select! recv，不再 100ms 轮询。
//! 同时保留 pollInbox 兼容语义与 JSON 持久化。

use crate::store::persistence;
use crate::store::types::*;
use parking_lot::Mutex;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::mpsc;

/// 收件箱消息（任务字条 或 通知字条）
#[derive(Debug, Clone)]
pub enum InboxItem {
    Task(TaskQueueItem),
    Notify(NotifyQueueItem),
}

/// 一个 group 的共享状态（任务队列 + 通知队列 + 各 agent 的 mpsc sender）
struct GroupState {
    tasks: Vec<TaskQueueItem>,
    notifies: Vec<NotifyQueueItem>,
    /// 每个 agent 的收件箱 sender（消息驱动推送）
    inboxes: HashMap<String, mpsc::UnboundedSender<InboxItem>>,
}

impl GroupState {
    fn new() -> Self {
        Self {
            tasks: vec![],
            notifies: vec![],
            inboxes: HashMap::new(),
        }
    }
}

/// 全局共享状态中心
#[derive(Default)]
pub struct SharedStateCenter {
    groups: Mutex<HashMap<String, GroupState>>,
}

static SHARED: once_cell::sync::Lazy<Arc<SharedStateCenter>> =
    once_cell::sync::Lazy::new(|| Arc::new(SharedStateCenter::default()));

pub fn shared_state() -> Arc<SharedStateCenter> {
    SHARED.clone()
}

impl SharedStateCenter {
    fn group<'a>(&'a self, map: &'a mut HashMap<String, GroupState>, group_id: &str) -> &'a mut GroupState {
        map.entry(group_id.to_string()).or_insert_with(GroupState::new)
    }

    /// 注册一个 agent 的收件箱 channel，返回 receiver
    pub fn register_inbox(
        &self,
        group_id: &str,
        agent_id: &str,
    ) -> mpsc::UnboundedReceiver<InboxItem> {
        let (tx, rx) = mpsc::unbounded_channel();
        {
            let mut map = self.groups.lock();
            let g = self.group(&mut map, group_id);
            g.inboxes.insert(agent_id.to_string(), tx);
        }
        rx
    }

    /// 注销收件箱
    pub fn unregister_inbox(&self, group_id: &str, agent_id: &str) {
        let mut map = self.groups.lock();
        if let Some(g) = map.get_mut(group_id) {
            g.inboxes.remove(agent_id);
        }
    }

    /// 向某 agent 收件箱「扔任务字条」，并唤醒其 engine
    pub fn push_task(&self, params: PushTaskParams) -> TaskQueueItem {
        let item = TaskQueueItem {
            id: crate::store::new_id("tq"),
            group_id: params.group_id,
            sender_id: params.sender_id,
            receiver_id: params.receiver_id,
            content: params.content,
            data: params.data,
            created_at: crate::store::now_iso(),
            status: "pending".into(),
            claimed_by: None,
            result: None,
            result_data: None,
            completed_at: None,
        };
        let inbox_tx;
        {
            let mut map = self.groups.lock();
            let g = self.group(&mut map, &item.group_id);
            g.tasks.push(item.clone());
            // 截断防止无限增长
            if g.tasks.len() > 2000 {
                let drop_n = g.tasks.len() - 2000;
                g.tasks.drain(0..drop_n);
            }
            inbox_tx = g.inboxes.get(&item.receiver_id).cloned();
        }
        self.persist(&item.group_id);
        if let Some(tx) = inbox_tx {
            let _ = tx.send(InboxItem::Task(item.clone()));
        }
        item
    }

    /// Agent 认领一个属于自己的 pending 任务
    pub fn claim_task(&self, group_id: &str, agent_id: &str, instance_id: &str) -> Option<TaskQueueItem> {
        let mut map = self.groups.lock();
        let g = self.group(&mut map, group_id);
        let idx = g
            .tasks
            .iter()
            .position(|t| t.receiver_id == agent_id && t.status == "pending")?;
        g.tasks[idx].status = "claimed".into();
        g.tasks[idx].claimed_by = Some(instance_id.to_string());
        let item = g.tasks[idx].clone();
        drop(map);
        self.persist(group_id);
        Some(item)
    }

    /// 完成任务并自动发布 task_complete/task_failed 通知
    pub fn complete_task(
        &self,
        task_id: &str,
        success: bool,
        result: Option<String>,
        result_data: Option<serde_json::Value>,
    ) -> Option<TaskQueueItem> {
        // 先在锁内找到并更新任务，提取通知所需字段
        let updated: Option<(TaskQueueItem, String, String)> = {
            let mut map = self.groups.lock();
            let mut found: Option<(TaskQueueItem, String, String)> = None;
            for g in map.values_mut() {
                if let Some(item) = g.tasks.iter_mut().find(|t| t.id == task_id) {
                    item.status = if success { "completed" } else { "failed" }.into();
                    item.result = result.clone();
                    item.result_data = result_data.clone();
                    item.completed_at = Some(crate::store::now_iso());
                    let notify_receiver = item.sender_id.clone();
                    let notify_sender = item.receiver_id.clone();
                    let task_item = item.clone();
                    found = Some((task_item, notify_sender, notify_receiver));
                    break;
                }
            }
            found
        };

        let (task_item, notify_sender, notify_receiver) = updated?;
        let group_id = task_item.group_id.clone();
        self.persist(&group_id);

        // 自动发布通知给原 sender（锁外执行，避免重入）
        self.push_notify(PushNotifyParams {
            group_id: group_id.clone(),
            kind: if success { "task_complete" } else { "task_failed" }.into(),
            sender_id: notify_sender,
            receiver_id: notify_receiver,
            content: result
                .clone()
                .unwrap_or_else(|| if success { "任务完成".into() } else { "任务失败".into() }),
            data: Some(serde_json::json!({
                "task_id": task_id,
                "extra": result_data.clone().unwrap_or(serde_json::Value::Null),
            })),
        });
        Some(task_item)
    }

    /// 列出某群的所有任务
    pub fn list_tasks(&self, group_id: &str) -> Vec<TaskQueueItem> {
        let mut map = self.groups.lock();
        self.group(&mut map, group_id).tasks.clone()
    }

    /// 向通知队列投递通知，并唤醒目标 agent
    pub fn push_notify(&self, params: PushNotifyParams) -> NotifyQueueItem {
        let item = NotifyQueueItem {
            id: crate::store::new_id("nq"),
            group_id: params.group_id,
            kind: params.kind,
            sender_id: params.sender_id,
            receiver_id: params.receiver_id,
            content: params.content,
            data: params.data,
            created_at: crate::store::now_iso(),
        };
        let inbox_tx;
        {
            let mut map = self.groups.lock();
            let g = self.group(&mut map, &item.group_id);
            g.notifies.push(item.clone());
            if g.notifies.len() > 500 {
                let drop_n = g.notifies.len() - 500;
                g.notifies.drain(0..drop_n);
            }
            // 定向通知：唤醒目标 agent；广播：唤醒群内所有 agent
            inbox_tx = if item.receiver_id == "broadcast" {
                // 收集所有 sender（clone 出来，避免持锁时 send）
                let senders: Vec<mpsc::UnboundedSender<InboxItem>> =
                    g.inboxes.values().cloned().collect();
                // 返回一个标记，外面逐个 send
                drop(map);
                for tx in senders {
                    let _ = tx.send(InboxItem::Notify(item.clone()));
                }
                None
            } else {
                g.inboxes.get(&item.receiver_id).cloned()
            };
        }
        self.persist(&item.group_id);
        if let Some(tx) = inbox_tx {
            let _ = tx.send(InboxItem::Notify(item.clone()));
        }
        item
    }

    /// 轮询收件箱（兼容旧 API）：返回属于该 agent 的 pending 任务 + 定向/广播通知
    pub fn poll_inbox(
        &self,
        group_id: &str,
        agent_id: &str,
        since: Option<&str>,
    ) -> (Vec<TaskQueueItem>, Vec<NotifyQueueItem>) {
        let mut map = self.groups.lock();
        let g = self.group(&mut map, group_id);
        let tasks: Vec<TaskQueueItem> = g
            .tasks
            .iter()
            .filter(|t| t.receiver_id == agent_id && t.status == "pending")
            .cloned()
            .collect();
        let notifies: Vec<NotifyQueueItem> = g
            .notifies
            .iter()
            .filter(|n| {
                if n.receiver_id != "broadcast" && n.receiver_id != agent_id {
                    return false;
                }
                match since {
                    Some(s) => n.created_at.as_str() > s,
                    None => true,
                }
            })
            .cloned()
            .collect();
        (tasks, notifies)
    }

    /// 获取群组快照
    pub fn snapshot(&self, group_id: &str) -> GroupQueueSnapshot {
        let mut map = self.groups.lock();
        let g = self.group(&mut map, group_id);
        GroupQueueSnapshot {
            group_id: group_id.to_string(),
            tasks: g.tasks.clone(),
            notifies: g.notifies.clone(),
        }
    }

    /// 从磁盘恢复某群的队列
    pub fn restore(&self, snap: GroupQueueSnapshot) {
        let mut map = self.groups.lock();
        let g = self.group(&mut map, &snap.group_id);
        g.tasks = snap.tasks;
        g.notifies = snap.notifies;
    }

    fn persist(&self, group_id: &str) {
        let snap = self.snapshot(group_id);
        persistence::schedule_save_queue(group_id, serde_json::to_value(&snap).unwrap());
    }
}

#[derive(Debug, Clone)]
pub struct PushTaskParams {
    pub group_id: String,
    pub sender_id: String,
    pub receiver_id: String,
    pub content: String,
    pub data: Option<serde_json::Value>,
}

#[derive(Debug, Clone)]
pub struct PushNotifyParams {
    pub group_id: String,
    pub kind: String,
    pub sender_id: String,
    pub receiver_id: String,
    pub content: String,
    pub data: Option<serde_json::Value>,
}

/// 从磁盘加载所有队列到内存（启动时）
pub fn load_all_queues() {
    let snapshots = persistence::load_all_queues();
    let ss = shared_state();
    let n = snapshots.len();
    for snap in snapshots {
        ss.restore(snap);
    }
    log::info!("[SharedState] restored {n} group queue(s)");
}
